File: baml_language/sdk_tests/crates/java/function_calls/customizable/TestMethodsOnClasses.java
// Static + instance method coverage (ns_methods_on_classes.Greeter).
//
// `raises_test.DocLoader` pins the *shape* of method bindings, but its bodies
// always throw, so they are never invoked. `Greeter` has non-throwing bodies,
// so this exercises the full host->engine round-trip for both flavors:
//   - static   -> `Greeter.create(name)`          (no receiver)
//   - instance -> `g.who()` / `g.greet(arg)`      (receiver is `self`)
// each with its `_async` sibling returning `CompletableFuture<T>`.
//
// Port of python_pydantic2/function_calls/customizable/test_methods_on_classes.py.

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertNotNull;

import baml_sdk.methods_on_classes.Greeter;
import java.util.concurrent.CompletableFuture;
import java.util.function.BiFunction;
import java.util.function.Function;
import org.junit.jupiter.api.Test;

class TestMethodsOnClasses {

    @Test
    void test_methods_on_classes_method_bindings_exist() {
        // java-port note: Python asserts `callable(Greeter.create)` etc. on the
        // (un)bound attributes. The Java analog is that each binding resolves as
        // a method reference at compile time (static bindings hang off the class;
        // instance bindings take the receiver as the unbound first parameter).
        Function<String, Greeter> create = Greeter::create;
        Function<String, CompletableFuture<Greeter>> createAsync = Greeter::create_async;
        Function<Greeter, String> who = Greeter::who;
        Function<Greeter, CompletableFuture<String>> whoAsync = Greeter::who_async;
        BiFunction<Greeter, String, String> greet = Greeter::greet;
        BiFunction<Greeter, String, CompletableFuture<String>> greetAsync = Greeter::greet_async;

        assertNotNull(create);
        assertNotNull(createAsync);
        assertNotNull(who);
        assertNotNull(whoAsync);
        assertNotNull(greet);
        assertNotNull(greetAsync);
    }

    @Test
    void test_methods_on_classes_static_create_round_trips() {
        Greeter g = Greeter.create("ada");
        assertInstanceOf(Greeter.class, g);
        assertEquals("ada", g.name());
    }

    @Test
    void test_methods_on_classes_static_create_async_round_trips() {
        Greeter g = Greeter.create_async("grace").join();
        assertInstanceOf(Greeter.class, g);
        assertEquals("grace", g.name());
    }

    @Test
    void test_methods_on_classes_instance_who_round_trips() {
        Greeter g = Greeter.create("hopper");
        assertEquals("hopper", g.who());
    }

    @Test
    void test_methods_on_classes_instance_who_async_round_trips() {
        Greeter g = Greeter.create_async("hopper").join();
        assertEquals("hopper", g.who_async().join());
    }

    @Test
    void test_methods_on_classes_instance_greet_with_arg_round_trips() {
        Greeter g = Greeter.create("lovelace");
        assertEquals("hi", g.greet("hi"));
    }

    @Test
    void test_methods_on_classes_instance_greet_async_with_arg_round_trips() {
        Greeter g = Greeter.create_async("lovelace").join();
        assertEquals("hi", g.greet_async("hi").join());
    }
}

File: baml_language/sdk_tests/crates/rust/function_calls/customizable/test_methods_on_classes.rs
//! Static + instance method coverage (ns_methods_on_classes.Greeter).
//!
//! `raises_test.DocLoader` pins the *shape* of method bindings, but its
//! bodies always throw, so they are never invoked. `Greeter` has non-throwing
//! bodies, so this exercises the full host->engine round-trip for both
//! flavors:
//!
//!   - static   -> `Greeter::create(name)`      (associated function, no `self`)
//!   - instance -> `g.who()` / `g.greet(arg)`   (`&self` receiver)
//!
//! each with its `_async` sibling; the async cases run under
//! `#[tokio::test]`.
//!
//! Node counterpart: `crates/typescript_node/function_calls/customizable/methods_on_classes.test.ts`.

use baml_sdk::methods_on_classes::Greeter;

#[test]
fn test_methods_on_classes_method_bindings_exist() {
    // Static bindings hang off the class; instance bindings carry `self`.
    // python asserts `callable(...)`; in Rust, naming each binding as a
    // function item is the (compile-time) assertion.
    let _ = (
        Greeter::create,
        Greeter::create_async,
        Greeter::who,
        Greeter::who_async,
        Greeter::greet,
        Greeter::greet_async,
    );
}

#[test]
fn test_methods_on_classes_static_create_round_trips() {
    // `isinstance(g, Greeter)` is the ascribed type.
    let g: Greeter = Greeter::create("ada".to_string()).unwrap();
    assert_eq!(g.name, "ada");
}

#[tokio::test]
async fn test_methods_on_classes_static_create_async_round_trips() {
    let g: Greeter = Greeter::create_async("grace".to_string()).await.unwrap();
    assert_eq!(g.name, "grace");
}

#[test]
fn test_methods_on_classes_instance_who_round_trips() {
    let g = Greeter::create("hopper".to_string()).unwrap();
    assert_eq!(g.who().unwrap(), "hopper");
}

#[tokio::test]
async fn test_methods_on_classes_instance_who_async_round_trips() {
    let g = Greeter::create_async("hopper".to_string()).await.unwrap();
    assert_eq!(g.who_async().await.unwrap(), "hopper");
}

#[test]
fn test_methods_on_classes_instance_greet_with_arg_round_trips() {
    let g = Greeter::create("lovelace".to_string()).unwrap();
    assert_eq!(g.greet("hi".to_string()).unwrap(), "hi");
}

#[tokio::test]
async fn test_methods_on_classes_instance_greet_async_with_arg_round_trips() {
    let g = Greeter::create_async("lovelace".to_string()).await.unwrap();
    assert_eq!(g.greet_async("hi".to_string()).await.unwrap(), "hi");
}
